### ISSUE TITLE
[ENHANCEMENT] Prevent add-ons from being disabled once in use

### ISSUE THREAD

**Issue by starbops** [harvester/harvester#6089]
**Is your enhancement related to a problem? Please describe.**

If there are any outstanding custom resources related to the add-on created, we should prevent users from blindly disabling the add-on. Otherwise, there will be dangling resources left in the system.

For example, suppose a user enabled the `harvester-vm-dhcp-controller` add-on and created a couple of IPPool and VirtualMachineNetworkConfig objects. In that case, these objects will be left in the system and become orphans once the user turns off the add-on.

The stuff left in the system might be more than just custom resources. Any transient and helper objects, such as pods, services, secrets, etc. that used to work together to provide functionality will also remain in the system, which might cause unexpected problems.

**Describe the solution you'd like**

Prevent in-use add-ons from being disabled with in-flight checks done by the validating admission webhook.

**Describe alternatives you've considered**

N/A

**Additional context**

This is somewhat of a chicken-and-egg problem: suppose an add-on has never been enabled before, which means no CRDs are installed. Then, it's impossible for the webhook to check if any "live" custom resources remain on the system. Plus, there are plenty of add-ons, all introducing different CRDs. There seems to be no place to hold the information for the webhook to know what kind of custom resources to check for what add-on.

**Comment by w13915984028** [comment]
Addons like `rancher-monitoring` and `rancher-logging`, have installed the CRD when cluster is built up via managedchart. The enable/disable of addon will not hit CRD existing/not-existing issue.

Disabling addon mainly removes the deployments to reduce system resource usage, but keeps CRD objects for re-using. Disable != delete addon. Sometimes disable & enable addon can also help troubleshooting & bug fix.

Preventing user from disable running addon seems more strict and not convenient in some cases.

Here we can have an `alternatives` to describe the behaviour in the document, if user wan to clean those remaining objects, they need to do it manually. Or we add an additional UI menu to allow user to clean the orphaned objects after addon is disabled.

**Comment by starbops** [comment]
@w13915984028 Thanks for the input. Indeed, there's not much we can do for those add-ons that we have less control over, e.g., `rancher-logging` and `rancher-monitoring` add-ons. As you said, the way to go is to provide documentation about how to clean up the remaining resources after disabling the add-on. But for our homemade add-ons like `vm-import-controller,` `vm-dhcp-controller`, etc., I think shipping them with a validating admission webhook, especially for their own Addon objects, is possible. The admission webhook validates the Addon objects' UPDATE requests, especially for the "disable" action. It will check whether any custom resources of its kind remain in the cluster. If that's the case, the webhook will reject the resource update.

For example, the validating admission webhook of the `vm-dhcp-controller` add-on will check for the existence of custom resources for the `harvester-vm-dhcp-controller` Addon object update requests.

In other words, it becomes the add-on authors' duty to implement such admission webhooks since they know best what resources the add-ons depend on.

What do you think?

**Comment by bk201** [comment]
@w13915984028 The scope here is only for add-ons that impact running instances.
- vm-dhcp-controller
- vm-import (when there is an import running)
- pci-device
- vGPU

Not to disable the addon in use is common sense, but we can improve the wrong operation by adding some checks in the webhooks.
We can skip logging and monitoring if they are not needed.

**Comment by w13915984028** [comment]
@bk201 @starbops 

That will make sense to add the check case by case. Some addons are running independently and some are as service for others.

Meanwhile, for each addon with this new webhook, we need detailed documents about how to step by step remove the related resource and finannly user can disable it. thanks.
